yt/utilities/io_handler.py

```diff
@@ -254,46 +196,29 @@ def _read_particle_selection(
             else:
                 ptf[ftype].append(fname)
                 field_maps[field].append(field)
-        # Now we have our full listing
-
-        # Now we add particle counts across chunks to psize
-        self._count_particles_chunks(psize, chunks, ptf, selector)
+            data[field] = []
 
-        # Now we allocate
-        for field in fields:
-            if field[0] in unions:
-                for pt in unions[field[0]]:
-                    fsize[field] += psize.get(pt, 0)
-            else:
-                fsize[field] += psize.get(field[0], 0)
-        shape: Tuple[int, ...]
-        for field in fields:
-            if field[1] in self._vector_fields:
-                vsize = self._vector_fields[field[1]]  # type:ignore
-                # note: the above line causes a mypy failure due to how we
-                # convert _vector_fields from a tuple to dict in __init__. mypy
-                # is expecting a tuple here. And since _vector_fields is used in
-                # many places, just ignoring for now...
-                shape = (fsize[field], vsize)
-            elif field[1] in self._array_fields:
-                shape = (fsize[field],) + self._array_fields[field[1]]
-            else:
-                shape = (fsize[field],)
-            rv[field] = np.empty(shape, dtype="float64")
-            ind[field] = 0
         # Now we read.
         for field_r, vals in self._read_particle_fields(chunks, ptf, selector):
             # Note that we now need to check the mappings
             for field_f in field_maps[field_r]:
-                my_ind = ind[field_f]
-                # mylog.debug("Filling %s from %s to %s with %s",
-                #    field_f, my_ind, my_ind+vals.shape[0], field_r)
-                rv[field_f][my_ind : my_ind + vals.shape[0], ...] = vals
-                ind[field_f] += vals.shape[0]
-        # Now we need to truncate all our fields, since we allow for
-        # over-estimating.
-        for field_f in ind:
-            rv[field_f] = rv[field_f][: ind[field_f]]
+                data[field_f].append(vals)
+
+        rv: Dict[Tuple[str, str], np.ndarray] = {}  # the return dictionary
+        fields = list(data.keys())
+        for field_f in fields:
+            # We need to ensure the arrays have the right shape if there are no
+            # particles in them.
+            total = sum(_.size for _ in data[field_f])
+            if total > 0:
+                rv[field_f] = np.concatenate(data.pop(field_f), axis=0)
```

Comment:
So the following should fix the failing tests (I only tested the gadget answer tests locally, but I suspect it'll fix the other failures too). The issue was the when we were pre-allocating we were setting the empty array to `dtype="float64"` and now we're not... and I guess the returns from all the various read functions for the individual frontends aren't totally consistent in what they return. So I'm inclined to just put the `astype` here, but are there downsides to that? Maybe there's a better way? 

```suggestion
                rv[field_f] = np.concatenate(data.pop(field_f), axis=0).astype("float64")
```

Oh and with numpy >= 1.20.0, concatenate actually accepts a dtype argument. It has a `casting` argument that could be useful: https://numpy.org/doc/stable/reference/generated/numpy.concatenate.html#numpy.concatenate But I didn't think it was worth adding a version dependent behavior here... but could?

Reply:
+ 1 for this patch

> But I didn't think it was worth adding a version dependent behavior here... but could?

FWIW, a _lot_ of functions in numpy acquired the `dtype` argument over the years, but the process is extremely gradual, so it's really hard to keep track of (and proabably not worth the trouble). In the future ™️, I envision that we could go through all usages of `.astype` manually and use `dtype` instead, but AFAIK there is'nt anything equivalent to e.g. pyupgrade that can be used to automate this process

Reply:
I'll give @matthewturk a chance to chime in on this before committing the change... I also considered doing the cast within the loop where we append to the `data` list, but that would necessitate dealing with empty chunks explicitly rather relying on concatenate to squash them.

Reply:
So I think my only concern with this is in the possible case of having an int array.  I do not actually remember if we support that in *general* for particle fields.  If we do not, then an astype should be fine.

Reply:
that's a good question... previously we were allocating all the arrays initially with `rv[field] = np.empty(shape, dtype="float64")` and then filling. That would have implicitly converted any int to float64, even if a reader returned an int array, ya? 

I think more broadly we could use a refactor to better handle particle dtypes. Or at least document how we handle them. The various `_read_` implementations for each frontend frequently include casting to float64 and it's never been clear to me when or exactly why it's needed. But I'm fairly certain that casting here is on par with what this function did previously.

Reply:
I think the dtype constraint really comes from yt's core, where many Cython functions expect float64 and nothing else. Though that's just an educated guess, I never tried looking deep into this.

Reply:
I'm going to go ahead and commit this suggestion...